CDATA codegen updates, part 3

silentbicycle opened the pull request:
Some further updates, for a new Device Detection canary. This still isn't quite ready to go upstream, but it's getting closer.

Changes:
- Move `.end` and `.endid_offset` fields from the state struct table to a separate table. Those fields aren't used until the end of input, moving them out makes the state table more dense and improves locality.
- Intern individual words from each state's 256-bitsets: These tend to be duplicated across several states, so if DFAs store them via four `uint8_t` or `uint16_t` offsets into a shared word table the overall binary ends up considerably smaller. They are sorted by use (descending), so the most frequently used ones are likely to stay in cache.
- Fix a memory leak -- already merged upstream, but has not been synced here: https://github.com/katef/libfsm/pull/504
- Buffer eager_outputs until a successful match, rather than immediately writing them into the caller's buffer.

silentbicycle committed:
cdata: Move .end and (optional) .endid_offset into a separata array.

These fields are only used at the end of input, and moving them into a
different struct will make the per-state data accessed during DFA
execution more compact.

silentbicycle committed:
cdata: Intern individual words from 256-bitsets.

Each state has two 256-bitsets, stored as a uint64_t[4], but the
individual words in those have a lot of duplication. Add a table
with every unique word, sorted descending by frequency, and replace
the per-state labels and label_group_starts arrays with an array of
offsets into the label_word table. Typically these offsets will
fit in a uint8_t (though the code generation will switch to
a uint16_t when necessary), making the per-state data much smaller.
The label_word table's most commonly used entries are all grouped
together and should stay in cache.

silentbicycle committed:
Fix memory leak.

The edge sets leak when halting with FSM_DETERMINISE_WITH_CONFIG_STATE_LIMIT_REACHED.

silentbicycle committed:
cdata: Buffer eager_outputs until match, rather than setting imediately.

Previously, the CDATA codegen wrote eager_outputs directly into the
caller's match bit buffer as they were encountered. Instead, set them in
a stack-allocated buffer, and then copy them to the caller's if the DFA
match succeeds overall.

In order to avoid repeatedly checking for whether an eager_output has
already been set (in the buffer), this collects the set of all distinct
eager_output IDs and then remaps the array with eager output IDs to
offsets into the unique set. This condenses the (sparse) set
into a dense series 0..n that can be represented by flags in a
stack-allocated bit vector (with a size known at compile time), and
redundant eager_outputs harmlessly set flag bits that are already set.
If the overall match succeeds, that bit vector is matched up with the
unique ID array and the sparse values are written into the caller's
buffer.

Because the unique ID array is sorted, the relative ordering of the
sparse and dense IDs is preserved (and 0 stays 0), so using
non-ascending values as terminators still works.

deg4uss3r reviewed (approved):
Admittedly I am coming into the code base pretty cold and C isn't my most preferred language but I do not see anything here that would prevent this from being approved.

silentbicycle committed:
Fix a bug in the CDATA interpreter's bitset checking.

Because of the for loop init condition, this was checking bits
`1 << 1..64` in every word -- it was unintentionally ignoring
the least significant bit. It should check `1 << 0..64`.

silentbicycle committed:
Rename test to _single, add _multiple variant.

These test combining `[""]` and `["", ""]`.

silentbicycle committed:
Fix memory leak in fsm_eager_output_compact, found while fuzzing.

silentbicycle committed:
Fix fsm_union_repeated_pattern_group's anchoring linkage.

Previously this didn't handle mixed anchoring correctly, potentially
leading to false positives the case represented by
eager_output_alt_mixing_anchored_and_unanchored.c.

See comments in fsm_union_repeated_pattern_group for details.

Fuzzing did not turn up any new issues.

Another commit after this will make a few small interface changes
and update callers.

silentbicycle committed:
fsm_union_repeated_pattern_group: Interface changes.

- Instead of taking an array of `struct fsm_union_array *` pointers,
  this now takes an array of `struct fsm *` pointers. The other
  fields on `fsm_union_array` are no longer used, so the extra
  struct layer has been removed.

- This now takes an extra argument, id_base, because each
  nfa[i] will get end IDs and/or output IDs (i + id_base) set on them.
  Previously these were set by the caller.

- Rename parameters, to emphasize that the FSMs must be NFAs.

- Update the test code for the interface changes.

- Remove flags from the test code that are no longer used.

silentbicycle committed:
fsm_union_repeated_pattern_group: fix linkage for mixed start anchoring.

deg4uss3r committed:
update CI to lock to ubuntu 22.04 for clang 14.0.0

deg4uss3r committed:
keep matrix.os so we can easily add more os variations in the future

silentbicycle committed:
Add tests, fix anchoring bugs in fsm_union_repeated_pattern_group.

There are two bugs captured in eager_output_unanchored_end_plus.c:

- Regexes ending in '+' weren't combining correctly, because analysis
  wasn't properly handling the construction for matching but optionally
  repeating the last character.

- Eager matching after consuming a single character from the start state
  wasn't linked correctly to the global_unanchored_start_loop, so while
  the labeled edges were copied the eager output was lost.

The other test files are focused on variants of that -- the + and
start cases individually, and when + precedes a `()` subtree with
more than one character.

silentbicycle committed:
Interface change: Add 'const'.

silentbicycle committed:
fuzz/target.c: fsm_union_repeated_pattern_group interface changes.

silentbicycle committed:
union: Fix trivial memory leak.

silentbicycle committed:
union.c: Add comments for assertions.

Fuzzing has produced inputs that cause this to fail, but they
all depend on embedded '\0' characters. I wasn't able to reproduce
the failure without those present, but I will investigate further
later. For now, adding a TODO.

silentbicycle committed:
Switch to collecting an anchored_start state set, not just one state.

`(^|wax-)((?:banana|^apple))` is an example of a regex that needs
multiple anchored_start states linked in order to combine correctly.

katef commented on src/libfsm/print/cdata.c:
```diff
@@ -1044,6 +1320,14 @@ populate_config_from_ir(struct cdata_config *config, const struct fsm_alloc *all
+	/* add a single entry for 0, in case the IR only has a single IR_NONE or IR_SAME state */
+	config->bitset_words.ceil = 1;
+	config->bitset_words.used = 1;
+	config->bitset_words.pairs = calloc(1, sizeof(config->bitset_words.pairs[0]));
```
missing error handling here, and also this should be `f_calloc()`

silentbicycle replied:
Fixed in a99b9137, now the caller also checks the return value and returns `-1` on error, as per the print interface.

silentbicycle committed:
cdata: Instead of asserting, the print interface returns -1 on error.

Bubble up the allocation failure errors.

Use `f_calloc` rather than `calloc`.

silentbicycle commented on Makefile:
```diff
@@ -41,6 +41,8 @@ CFLAGS += -DEXPENSIVE_CHECKS
 DEBUG ?= 1
 .endif
 
+CFLAGS += -DNOSTRIP
```
According to 5df68c8150c0785550dd0987bb9f14efa5d36acb it was added temporarily while investigating the AUSAN CI issue. Removed.

silentbicycle committed:
cdata: Bubble up other allocation errors.

silentbicycle commented on src/libfsm/print/cdata.c:
```diff
@@ -884,6 +1118,42 @@ append_dst(const struct fsm_alloc *alloc, struct dst_buf *buf, uint32_t dst)
+		const size_t nceil = (bws->ceil == 0 ? 8 : 2*bws->ceil);
+		struct bitset_word_pair *npairs = f_realloc(alloc,
+		    bws->pairs, nceil * sizeof(npairs[0]));
+		assert(npairs != NULL);
```
Fixed in 05096968.

silentbicycle committed:
retest: extra error checking for retest's fileno, isatty

Something strange is happening in CI, and these may be involved.

silentbicycle committed:
cdata: change return types to void, since these functions can't fail.

silentbicycle committed:
cdata: Error check.

silentbicycle committed:
Merge pull request #34 from fastly/sv/fix-anchoring-for-fsm_union_repeated_pattern_group

Fix anchoring for `fsm_union_repeated_pattern_group`

silentbicycle committed:
cdata: Fix NULL check. This looks like a merge error.

silentbicycle committed:
github workflows: ensure ${env.wc} exists.

silentbicycle committed:
Merge branch 'main' into sv/cdata-codegen-updates-part-3

silentbicycle committed:
cdata: Add a few comments about internal structs.

silentbicycle commented:
Force pushing to drop several temporary commits investigating CI failures (ultimately due to https://www.githubstatus.com/incidents/9qcwpy3ckdrf)

silentbicycle committed:
Revert "github workflows: ensure ${env.wc} exists."

This reverts commit 6078cdff247a7618e9e1bb4006dfc238af9da4ee.

This is not the right place to do this, and this wasn't the actual
root cause of the CI failures.
